src/common/snippets/src/runtime_configurator.cpp

```diff
@@ -57,18 +44,34 @@ void RuntimeConfigurator::initialization(const lowered::LinearIRCPtr& linear_ir)
     m_latest_shapes.resize(m_io_num);
     m_config->io_data_offsets.resize(m_io_num);
     m_config->tile_rank = linear_ir->get_config().m_loop_depth;
+    m_optimizer.init(linear_ir, m_io_descs, m_in_num);
 }
 
 void RuntimeConfigurator::update(const lowered::LinearIRCPtr& linear_ir) {
+    m_config->master_shape = linear_ir->get_master_shape();
+
+    LoopInfoRuntimeParamsMap initialized_info;
+    auto shapes = extract_shapes();
+    auto layouts = extract_layouts();
+    if (m_optimizer.enabled()) {
+        m_optimizer.optimize(m_config->master_shape, initialized_info, shapes, layouts, m_in_num);
+        update_tensor_rank(m_config->master_shape);
+    }
+
     if (linear_ir->is_dynamic()) {
-        update_loop_info(linear_ir);
+        update_loop_info(linear_ir, initialized_info);
```

Comment:
Can we make the optimizer responsible for calling `update_loop_info` for affected loops?
For example:
1) Implement `update_loop_info(loop_id)` that updates 1 loop
2) modify `update_loop_info(linear_ir)` so that it just calls `update_loop_info(loop_id)` for all loops in loop_map
3) Update loops before M dim optimization, and make the optimizer to update the necessary loops if the dimension was splitted.
This would enchance modularity and would simplify the logic. Otherwise we have to complicate signatures, and carry some opaque states from optimizer downwards (shapes/layouts/loop_info), and we don't even know if the states were updated or not.

Reply:
In general, we should think of the optimizer as a pass: it takes lir and updates everything that needs to be updated, so all other passes/functions could work completely independently (not semantically, since passes may depend on each other, but functionally, i.e. the only state that is passes between the passes (sorry 🙃) is lir itself

Reply:
Again, `update_loop_info` is `RuntimeConfigurator`'s API, so it should be extracted to apply your suggestion. It seems like the best way here is to extract it as a separate lowered pass. I believe this should be done within work on 148891 ticket (where all `RuntimeConfigurator::update` work should be done by passes) in order to avoid a mess of 2 approaches in one place

Reply:
The optimizer is currently implemented as a part of RuntimeConfigurator, so as a quick solution we can pass the configurator to the optimizer to make the necessary updates. The idea is to isolate all the optimization-related logic in one class.
Yes, the pass solution is better, and it's a long-term fix, but it's not clear when it will be addressed. So I would still suggest to update the optimizer a bit now, if it won't take too much time.

Reply:
The refactoring is done in the follow-up PR